src/databricks/labs/ucx/mixins/cached_workspace_path.py

```diff
@@ -1,139 +1,138 @@
-        path = str(cached_path)
-        if path in self._datas:
-            self._datas.move_to_end(path)
-            return _CachedIO(self._datas[path]).with_mode(mode)
-        io_obj = WorkspacePath.open(cached_path, mode, buffering, encoding, errors, newline)
-        # can't read twice from an IO so need to cache data rather than the io object
-        data = io_obj.read()
-        self._datas[path] = data
-        result = _CachedIO(data).with_mode(mode)
-        if len(self._datas) > self._max_entries:
+        # Need to bypass the _CachedPath.open() override to actually open and retrieve the file content.
+        with WorkspacePath.open(cached_path, mode="rb") as workspace_file:
+            data = workspace_file.read()
+        if self._max_entries <= len(self._datas):
             self._datas.popitem(last=False)
-        return result
+        self._datas[normalized_path] = data
+        return data
 
-    def clear(self):
+    def clear(self) -> None:
         self._datas.clear()
 
-    def remove(self, path: str):
-        if path in self._datas:
-            self._datas.pop(path)
+    def remove(self, path: _CachedPath) -> None:
+        del self._datas[self._normalize(path)]
 
 
 class _CachedPath(WorkspacePath):
-    def __init__(self, cache: _PathLruCache, ws: WorkspaceClient, *args: str | bytes | os.PathLike):
+    def __init__(self, cache: _PathLruCache, ws: WorkspaceClient, *args: str | bytes | os.PathLike) -> None:
         super().__init__(ws, *args)
         self._cache = cache
 
-    def with_object_info(self, object_info: ObjectInfo):
-        self._cached_object_info = object_info
-        return self
-
-    def with_segments(self, *path_segments: bytes | str | os.PathLike) -> _CachedPath:
+    @classmethod
+    def _from_object_info_with_cache(
+        cls,
+        cache: _PathLruCache,
+        ws: WorkspaceClient,
+        object_info: ObjectInfo,
+    ) -> _CachedPath:
+        assert object_info.path
+        path = cls(cache, ws, object_info.path)
+        path._cached_object_info = object_info
+        return path
+
+    def with_segments(self: _CachedPathT, *path_segments: bytes | str | os.PathLike) -> _CachedPathT:
         return type(self)(self._cache, self._ws, *path_segments)
 
     def iterdir(self) -> Generator[_CachedPath, None, None]:
+        # Variant of the superclass implementation that preserves the cache, as well as the client.
         for object_info in self._ws.workspace.list(self.as_posix()):
-            path = object_info.path
-            if path is None:
-                msg = f"Cannot initialise without object path: {object_info}"
-                raise ValueError(msg)
-            child = _CachedPath(self._cache, self._ws, path)
-            yield child.with_object_info(object_info)
-
-    def open(
+            yield self._from_object_info_with_cache(self._cache, self._ws, object_info)
+
+    @classmethod
+    def _detect_encoding_bom(cls, data: io.BytesIO) -> str | None:
```

Comment:
dupe of `databricks.labs.ucx.source_code.base.guess_encoding`.

Reply:
Refactored, and updated it at the same time to avoid double-opening files.